### PR TITLE
Integrate measureme's hardware performance counter support.

### PR DESCRIPTION
**Pull request by eddyb**
*Note: this is a companion to https://github.com/rust-lang/measureme/pull/143, and duplicates some information with it for convenience*

**(much later) EDIT**: take any numbers with a grain of salt, they may have changed since initial PR open.

## Credits

I'd like to start by thanking @alyssais, @cuviper, @edef1c, @glandium, @jix, @Mark-Simulacrum, @m-ou-se, @mystor, @nagisa, @puckipedia, and @yorickvP, for all of their help with testing, and valuable insight and suggestions.
Getting here wouldn't have been possible without you!

(If I've forgotten anyone please let me know, I'm going off memory here, plus some discussion logs)

## Summary

This PR adds support to `-Z self-profile` for counting hardware events such as "instructions retired" (as opposed to being limited to time measurements), using the `rdpmc` instruction on `x86_64` Linux.

While other OSes may eventually be supported, preliminary research suggests some kind of kernel extension/driver is required to enable this, whereas on Linux any user can profile (at least) their own threads.

Supporting Linux on architectures other than x86_64 should be much easier (provided the hardware supports such performance counters), and was mostly not done due to a lack of readily available test hardware.
That said, 32-bit `x86` (aka `i686`) would be almost trivial to add and test once we land the initial `x86_64` version (as all the CPU detection code can be reused).

A new flag `-Z self-profile-counter` was added, to control which of the named `measureme` counters is used, and which defaults to `wall-time`, in order to keep `-Z self-profile`'s current functionality unchanged (at least for now).

The named counters so far are:
* `wall-time`: the existing time measurement
    * name chosen for consistency with `perf.rust-lang.org`
    * continues to use `std::time::Instant` for a nanosecond-precision "monotonic clock"
* `instructions:u`: the hardware performance counter usually referred to as "Instructions retired"
    * here "retired" (roughly) means "fully executed"
    * the `:u` suffix is from the Linux `perf` tool and indicates the counter only runs while userspace code is executing, and therefore counts no kernel instructions
        * *see [Caveats/Subtracting IRQs](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Subtracting-IRQs) for why this isn't entirely true and why `instructions-minus-irqs:u` should be preferred instead*
* `instructions-minus-irqs:u`: same as `instructions:u`, except the count of hardware interrupts ("IRQs" here for brevity) is subtracted
    * *see [Caveats/Subtracting IRQs](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Subtracting-IRQs) for why this should be preferred over `instructions:u`*
* `instructions-minus-r0420:u`: experimental counter, same as `instructions-minus-irqs:u` but subtracting an undocumented counter (`r0420:u`) instead of IRQs
    * the `rXXXX` notation is again from Linux `perf`, and indicates a "raw" counter, with a hex representation of the low-level counter configuration - this was picked because we still don't *really* know what it is
    * this only exists for (future) testing and isn't included/used in any comparisons/data we've put together so far
    * *see [Challenges/Zen's undocumented 420 counter](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Epilogue-Zen’s-undocumented-420-counter) for details on how this counter was found and what it does*

---

There are also some additional commits:
* ~~see [Challenges/Rebasing *shouldn't* affect the results, right?](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Rebasing-*shouldn’t*-affect-the-results,-right) for details on the changes to `rustc_parse` and `rustc_trait_section` (the latter far more dubious, and probably shouldn't be merged, or not as-is)~~
  *  **EDIT**: the effects of these are no long quantifiable, the PR includes reverts for them
* ~~see [Challenges/`jemalloc`: purging will commence in ten seconds](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#jemalloc-purging-will-commence-in-ten-seconds) for details on the `jemalloc` change~~
  * this is also separately found in #77162, and we probably want to avoid doing it by default, ideally we'd use the runtime control API `jemalloc` offers (assuming that can stop the timer that's already running, which I'm not sure about)
  * **EDIT**: until we can do this based on `-Z` flags, this commit has also been reverted
* the `proc_macro` change was to avoid randomized hashing and therefore ASLR-like effects

---

**(much later) EDIT**: take any numbers with a grain of salt, they may have changed since initial PR open.

#### Write-up / report

Because of how extensive the full report ended up being, I've kept most of it [on `hackmd.io`](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view), but for convenient access, here are all the sections (with individual links):
<sup>(someone suggested I'd make a backup, so [here it is on the wayback machine](http://web.archive.org/web/20201127164748/https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view) - I'll need to remember to update that if I have to edit the write-up)</sup>

* [**Motivation**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Motivation)

* [**Results**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Results)
    * [**Overhead**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Overhead)
    *Preview (see the report itself for more details):*

    |Counter|Total<br>`instructions-minus-irqs:u`|Overhead from "Baseline"<br>(for all 1903881<br>counter reads)|Overhead from "Baseline"<br>(per each counter read)|
    |-|-|-|-|
    |Baseline|63637621286 ±6||
    |`instructions:u`|63658815885 ±2|&nbsp;&nbsp;+21194599 ±8|&nbsp;&nbsp;+11|
    |`instructions-minus-irqs:u`|63680307361 ±13|&nbsp;&nbsp;+42686075 ±19|&nbsp;&nbsp;+22|
    |`wall-time`|63951958376 ±10275|+314337090 ±10281|+165|

    * [**"Macro" noise (self time)**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#“Macro”-noise-(self-time))
    *Preview (see the report itself for more details):*

    || `wall-time` (ns) | `instructions:u` | `instructions-minus-irqs:u`
    -: | -: | -: | -:
    `typeck` | 5478261360 ±283933373 (±~5.2%) | 17350144522 ±6392 (±~0.00004%) | 17351035832.5 ±4.5 (±~0.00000003%)
    `expand_crate` | 2342096719 ±110465856 (±~4.7%) | 8263777916 ±2937 (±~0.00004%) | 8263708389 ±0 (±~0%)
    `mir_borrowck` | 2216149671 ±119458444 (±~5.4%) | 8340920100 ±2794 (±~0.00003%) | 8341613983.5 ±2.5 (±~0.00000003%)
    `mir_built` | 1269059734 ±91514604 (±~7.2%) | 4454959122 ±1618 (±~0.00004%) | 4455303811 ±1 (±~0.00000002%)
    `resolve_crate` | 942154987.5 ±53068423.5 (±~5.6%) | 3951197709 ±39 (±~0.000001%) | 3951196865 ±0 (±~0%)

    * [**"Micro" noise (individual sampling intervals)**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#“Micro”-noise-(individual-sampling-intervals))

* [**Caveats**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Caveats)
    * [**Disabling ASLR**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Disabling-ASLR)
    * [**Non-deterministic proc macros**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Non-deterministic-proc-macros)
    * [**Subtracting IRQs**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Subtracting-IRQs)
    * [**Lack of support for multiple threads**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Lack-of-support-for-multiple-threads)

* [**Challenges**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Challenges)
    * [**How do we even read hardware performance counters?**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#How-do-we-even-read-hardware-performance-counters)
    * [**ASLR: it's free entropy**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#ASLR-it’s-free-entropy)
    * [**The serializing instruction**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#The-serializing-instruction)
    * [**Getting constantly interrupted**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Getting-constantly-interrupted)
    * [**AMD patented time-travel and dubbed it `SpecLockMap`<br><sup>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;or: "how we accidentally unlocked `rr` on AMD Zen"</sup>**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#AMD-patented-time-travel-and-dubbed-it-SpecLockMapnbspnbspnbspnbspnbspnbspnbspnbspor-“how-we-accidentally-unlocked-rr-on-AMD-Zen”)
    * [**`jemalloc`: purging will commence in ten seconds**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#jemalloc-purging-will-commence-in-ten-seconds)
    * [**Rebasing *shouldn't* affect the results, right?**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Rebasing-*shouldn’t*-affect-the-results,-right)
    * [**Epilogue: Zen's undocumented 420 counter**](https://hackmd.io/sH315lO2RuicY-SEt7ynGA?view#Epilogue-Zen’s-undocumented-420-counter)

### CODE REVIEW

**Comment by eddyb** [comment]
cc @Amanieu Looks like we generate LLVM inline assembly that's not compatible with older versions? (i.e. LLVM 8 on CI)

**Comment by bjorn3** [comment]
Intel syntax is not supported by older versions of LLVM. Only AT&T syntax is.

**Comment by bjorn3** [comment]
Regressions of up to 3%.

**Comment by eddyb** [comment]
@Mark-Simulacrum Just to be sure, do we run the `-Z self-profile` for the query info under `perf stat`, and if so, do we mix those results into the non-`-Z self-profile`? Because `-Z self-profile` *should* be slightly slower, as we have to `match` on the `Counter`.

**Comment by Mark-Simulacrum** [comment]
Yes, but we take the minimum of all runs under perf stat (including the self profile one), and always have at least 2 (and IIRC at most 3 right now) runs, including one self profile run.

**Comment by Amanieu** [comment]
> Intel syntax is not supported by older versions of LLVM. Only AT&T syntax is.

You can use `options(att_syntax)` and then use AT&T syntax for the assembly code and it will work on all LLVM versions.

**Review by davidtwco — APPROVED** [review]
Implementation LGTM, r=me when ready

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (eddyb) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1315]: @davidtwco @nikomatsakis Just it doesn't get lost, I don't think we should merge with this, but I also am not sure what to do here. Do you think it would be reasonable to limit this to when `-Z self-profile` is enabled?

It's also not easy to test the impact of this fence.

**Comment by eddyb** [comment]
Given #78785, it might make sense to define `getrandom` in `librustc_driver` and have new flag for replacing `getrandom` with a deterministic PRNG (e.g. `-Z getrandom-seed`?) - note that this would not only affect proc macros (e.g. serde) using their own `HashMap`s, but also uses of randomness in the `const_random` proc macro, or for things like temp dir names in `rustc` itself.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (davidtwco) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:1315]: I'm not too sure either. I personally wouldn't have a problem with this being limited to when `-Z self-profile` is enabled.

Comment (bjorn3) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:1315]: Checking for `-Zself-profile` is probably more expensive than unconditionally running `mfence`.

**Comment by eddyb** [comment]
> Intel syntax is not supported by older versions of LLVM. Only AT&T syntax is.

I was confused about this, because I could've sworn I've used Intel syntax inline `asm!` *many years* ago.
LLVM documents `inteldialect` since 3.2: https://releases.llvm.org/3.2/docs/LangRef.html#inlineasm

The code emitting the error we're seeing is [still around](https://github.com/llvm/llvm-project/blob/95956c1c9aae7ea21c2b2f7a21e0901d549bd190/llvm/lib/CodeGen/AsmPrinter/AsmPrinterInlineAsm.cpp#L240-L242), but the `${` handling looks like it might be new.

And, there it is: operand modifiers for Intel syntax is new in LLVM 10 (https://github.com/llvm/llvm-project/commit/dc5b614fa9a1c83e8275fcb9c3f78444d0a30514)

So the problem is the `{...:e}` modifiers in order to get 32-bit registers instead of the 64-bit ones.
Oh well, I guess I am going with AT&T syntax after all (after checking that the generated assembly is identical).

**Comment by eddyb** [comment]
Looks like the CI build got much farther this time (before getting cancelled by the check failure), so that worked.

**Comment by crlf0710** [comment]
Triage: there's merge conflicts.

**Comment by eddyb** [comment]
My bad, this never landed because it needed more eyes I think? (tho @davidtwco did approve it, so I could've landed it based on that?) - and I ended up too swamped with other work to keep track of `rust-lang/rust` stuff.

**Comment by eddyb** [comment]
Looks like this is blocked on https://github.com/rust-lang/measureme/pull/147 landing and getting released, which was blocked on the CI success in https://github.com/rust-lang/rust/pull/78781#issuecomment-731944971 - but that's all been resolved, it just wasn't communicated properly (and then I dropped off).

**Comment by oli-obk** [comment]
r? @oli-obk 

stealing the PR as I also want to do a review round

File: compiler/rustc_trait_selection/src/traits/select/mod.rs
Comment (oli-obk) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:1315]: I'm fine leaving it in as it is for now. We can analyze its impact once perftest multithreaded rustc

**Comment by bstrie** [comment]
> Looks like this is blocked on rust-lang/measureme#147 landing and getting released

I see that this has landed. Is this PR unblocked?

File: src/tools/tidy/src/extdeps.rs
Comment (oli-obk) [inline src/tools/tidy/src/extdeps.rs:10]: This needs undoing now, other than that, the PR lgtm

**Comment by JohnCSimon** [comment]
Triage: 
@eddyb merge conflicts

**Comment by eddyb** [comment]
I am rebasing this right now, after https://github.com/rust-lang/measureme/pull/185 reminding me of it.

<sup>(If anyone is wondering why this stalled - it was part of a grant that we had to back out of because the goal had become unreachable/irrelevant due to unrelated improvements in `rustc`, and have been busy with other things)</sup>

**Comment by eddyb** [comment]
Blockers to merging:
* waiting on `measureme` `10.0.0` being released (cc @michaelwoerister @wesleywiser)
* as per #77162 the `jemalloc` workaround is most likely a slowdown
  * will check if perf is still affected, on this PR
  * it's still needed, with `jemalloc = true` in `config.toml` (as official releases are)
  * not sure yet if there's a way to dynamically set it based on `-Z` flags
* probably not a blocker, but worth mentioning it anyway: we could hijack `getrandom` to force `std` `HashMap` into deterministic behavior, with something like this (but ideally dynamically controlled by flags):
  ```rust
  #[no_mangle]
  pub unsafe extern "C" fn getrandom(p: *mut u8, n: usize, _: u32) -> usize {
      p.write_bytes(0, n);
      n
  }
  ```

**EDIT**: apparently https://github.com/rust-lang/measureme/pull/185 was wrong about LLVM 10 being enough to switch to Intel syntax? (see CI failure above)

**Comment by eddyb** [comment]
That doesn't look right, is perf broken currently? (Zero benchmarks - and clearly a regression from the bootstrap timings)
cc @Mark-Simulacrum

**Comment by eddyb** [comment]
Now using the just-released `measureme` `10.0.0` - thanks, @michaelwoerister!

@bors try @rust-timer queue

**Comment by Mark-Simulacrum** [comment]
Ryan and I are meeting this morning on the upgrade of perf to latest summarize, but until then I suspect this won't work:

```
 stderr=    Checking issue-32278-big-array-of-strings v0.1.0 (/tmp/.tmpqfgmrA)                                                                                                              +
 thread 'main' panicked at 'couldn't read string_data file: Os { code: 2, kind: NotFound, message: "No such file or directory" }', analyzeme/src/profiling_data.rs:50:27                    +
 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace                                                                                                              +
 Error: "Error reading top-level stream in file `issue_32278_big_array_of_strings-0016902.mm_profdata`: Expected file format version 7 but found `8`"                                       +
 thread 'main' panicked at 'failed to run summarize and summarize-9.0. Errors:                                                                                                              +
 summarize: failed to run summarize in "/tmp/.tmpqfgmrA/self-profile-output" with prefix "issue_32278_big_array_of_strings-0016902.mm_profdata"                                             +
 summarize-9.0: failed to run summarize-9.0 in "/tmp/.tmpqfgmrA/self-profile-output" with prefix "issue_32278_big_array_of_strings-0016902.mm_profdata"', collector/src/rustc-fake.rs:373:21+
 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace                                                                                                              +
 error: could not compile `issue-32278-big-array-of-strings`                                                                                                                                +
```

**Comment by eddyb** [comment]
Ahh that makes sense, I forgot version 10 is required to read version-10-generated profiles, my bad.

**Comment by eddyb** [comment]
In the interest of merging this PR sooner than later, I've also pushed a revert for the `jemalloc` "no purge" commit - however, I can confirm that the env var also works, i.e. the following has the same effect:
```sh
MALLOC_CONF='dirty_decay_ms:0,muzzy_decay_ms:0' setarch -R ./build/.../stage1/bin/rustc ...
```
(where `setarch -R` turns ASLR off - AFAIK, perf.rust-lang.org also does this)

However, the perf.rust-lang.org numbers will be discontinuous if that env var is suddenly set, and it also is only useful for runs that exceed 10 seconds of compile time (otherwise the timer never expires).
So its place is probably more in a "bag of tricks you'll need to make full use of `-Z self-profile`".

**Review by oli-obk — CHANGES_REQUESTED** [review]
just the logging nit, then r=me

**Comment by camelid** [comment]
triage: @eddyb could you rebase this PR?

**Comment by Dylan-DPC** [comment]
Closing this as inactive

**Comment by eddyb** [comment]
> We should be on latest measureme/analyzeme now on rustc-perf -- @eddyb can you rebase this and give it another bors try + rust-timer queue?

Oops, I think I missed this when it got fixed and forgot all the reverts postponing non-determinism fixes and whatnot, had already been handled. Rebasing now, hopefully for the last time.

**Comment by oli-obk** [comment]
The cargo regression shows up in cachegrind as

```
43,283,638  ???:<proc_macro::bridge::server::Dispatcher<proc_macro::bridge::server::MarkedTypes<rustc_expand::proc_macro_server::Rustc>> as proc_macro::bridge::server::DispatcherTrait>::dispatch
 6,527,962  ???:<hashbrown::raw::RawTable<(proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, core::num::nonzero::NonZeroU32)>>::reserve_rehash::<hashbrown::map::make_hasher<proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, core::num::nonzero::NonZeroU32, proc_macro::bridge::handle::NonRandomState>::{closure
-4,330,546  ???:<alloc::vec::Vec<(rustc_ast::tokenstream::TokenTree, rustc_ast::tokenstream::Spacing)> as core::ops::drop::Drop>::drop
 3,218,126  ???:<alloc::rc::Rc<alloc::vec::Vec<(rustc_ast::tokenstream::TokenTree, rustc_ast::tokenstream::Spacing)>> as core::ops::drop::Drop>::drop
-3,170,733  ???:<hashbrown::raw::RawTable<(proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, core::num::nonzero::NonZeroU32)>>::reserve_rehash::<hashbrown::map::make_hasher<proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Ident, proc_macro::bridge::client::Ident>, core::num::nonzero::NonZeroU32, std::collections::hash::map::RandomState>::{closure
 2,820,054  library/std/src/collections/hash/map.rs:std::collections::hash::map::DefaultHasher::new
 2,540,092  ???:<proc_macro::bridge::handle::NonRandomState as core::hash::BuildHasher>::hash_one::<&proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Punct, proc_macro::bridge::client::Punct>>
 1,880,036  library/core/src/hash/sip.rs:std::collections::hash::map::DefaultHasher::new
 1,655,610  ???:<proc_macro::bridge::handle::NonRandomState as core::hash::BuildHasher>::hash_one::<&proc_macro::bridge::Marked<rustc_span::span_encoding::Span, proc_macro::bridge::client::Span>>
 1,575,228  ???:<core::result::Result<proc_macro::bridge::Marked<rustc_span::span_encoding::Span, proc_macro::bridge::client::Span>, proc_macro::bridge::rpc::PanicMessage> as proc_macro::bridge::rpc::Encode<proc_macro::bridge::client::HandleStore<proc_macro::bridge::server::MarkedTypes<rustc_expand::proc_macro_server::Rustc>>>>::encode
-1,492,270  ???:<hashbrown::raw::RawTable<(proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Punct, proc_macro::bridge::client::Punct>, core::num::nonzero::NonZeroU32)>>::reserve_rehash::<hashbrown::map::make_hasher<proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Punct, proc_macro::bridge::client::Punct>, proc_macro::bridge::Marked<rustc_expand::proc_macro_server::Punct, proc_macro::bridge::client::Punct>, core::num::nonzero::NonZeroU32, std::collections::hash::map::RandomState>::{closure
```

the token stress test shows up as

```
2,966,103  ???:<proc_macro::bridge::server::Dispatcher<proc_macro::bridge::server::MarkedTypes<rustc_expand::proc_macro_server::Rustc>> as proc_macro::bridge::server::DispatcherTrait>::dispatch
```

Not sure why the proc macro bridge would be affected by this PR, but the cargo example seems to suggest the deterministic hasher changes by themselves also have some sort of an impact.

**Comment by eddyb** [comment]
Found likely culprit, missing `#[inline]` on these: https://github.com/rust-lang/rust/blob/083721a1a7365d3afe1521cd2661b2201aac0450/library/std/src/collections/hash/map.rs#L3136-L3159

Don't think anyone uses them so I won't bother with a separate benchmarking PR, just add a fix to this PR directly.

**Comment by eddyb** [comment]
@oli-obk Should be ready to merge now, the remaining `tt-muncher` thing feels like noise or getting unlucky with the choice of deterministic hash key.

File: compiler/rustc_data_structures/src/profiling.rs
Comment (bjorn3) [inline compiler/rustc_data_structures/src/profiling.rs:561]: *deterministic

**Comment by eddyb** [comment]
Oops this is what I get for not testing it, only realized it now that this is actually broken:
```console
$ echo 'fn main() {}' | rustc +nightly - -Z self-profile -Z self-profile-counter=instructions-minus-irqs:u
warning: failed to create profiler: only supported with measureme's "nightly" feature
```

I removed `features = ["nightly"]` because I thought that got removed from `measureme`, but I guess that's in a newer version than what `rustc` currently uses.

**Comment by eddyb** [comment]
```console
$ echo 'fn main() {}' | rustc +nightly-2022-07-23 - -Z self-profile -Z self-profile-counter=instructions-minus-irqs:u
warning: failed to create profiler: only supported with measureme's "nightly" feature

$ echo 'fn main() {}' | rustc +nightly-2022-07-24 - -Z self-profile -Z self-profile-counter=instructions-minus-irqs:u
thread 'opt hkqbnt908e7l8ng' panicked at 'assertion failed: end <= MAX_INTERVAL_VALUE', /cargo/registry/src/github.com-1ecc6299db9ec823/measureme-10.1.0/src/raw_event.rs:56:9
```

@jyn514 mentioned it never worked, to me, and... yeah, we added `features = ["nightly"]` but ~~`measureme 10.*` is just broken, unlike `measureme 9.*`?~~ (or did I never test multi-thread?)

---

So this is the `cargo check` equivalent (which is most of how I tested the counters):
```console
$ echo 'fn main() {}' | rustc +nightly-2022-07-24 - -Z self-profile -Z self-profile-counter=instructions-minus-irqs:u --emit=metadata
$ summarize summarize unknown-crate-*.mm_profdata
Segmentation fault (core dumped)
```

That's a segfault in `prettytable::TableSlice::get_all_column_width`?? `summarize summarize --json` works tho.

**Pull request merged**
